### PR TITLE
Call the correct deleter for ReadWriteVector::values

### PR DESCRIPTION
**Pull request by masterleinad**
The memory for `ReadWriteVector::values` is allocated in `Utilities::System::posix_memalign` (https://github.com/dealii/dealii/blob/master/include/deal.II/lac/read_write_vector.templates.h#L63)
which in turn either calls `::posix_memalign` or `malloc` (https://github.com/dealii/dealii/blob/master/source/base/utilities.cc#L731). Hence, the memory is allocated by `malloc` in the end and the correct deleter is `free` and not `delete[]`.
Detected running with `-fsanitize-address`.

### CODE REVIEW

File: include/deal.II/lac/read_write_vector.h [hunk 1]
```diff
@@ -610,7 +610,7 @@ namespace LinearAlgebra
     /**
      * Pointer to the array of local elements of this vector.
      */
-    std::unique_ptr<Number[]> values;
+    std::unique_ptr<Number[], decltype (free) *> values = std::unique_ptr<Number[], decltype (free) *> (nullptr, free);
```
Comment (tamiko) [inline include/deal.II/lac/read_write_vector.h]: isn't the type already complete determined by the rvalue/xvalue? I.e. can we use
```
auto values = std::unique_ptr(nullptr, free);
```
instead?

Comment (tamiko) [reply include/deal.II/lac/read_write_vector.h]: Ah, this is a member declaration. Sorry for the confusion.
At least a while ago we decided to not use the copy assignment operator for member initialization. Is this still the case?
What about you initialize with `values(nullptr, free)` in the constructor?

**Comment by drwells** [comment]
@tamiko I believe you approved this before the most recent patch. Is this okay? If so lets merge.

**Pull request merged**